### PR TITLE
konduit.yaml RFC (basic implementation)

### PR DESCRIPTION
**Pull request by maxpumperla**
# Konduit CLI for quick Python experiments

Install with `python setup.py install` from this folder. This will expose a CLI tool called
`konduit`. Assuming you have put a `konduit.jar` in the `tests` folder as described
in the main `README.md`, you can test this tool as follows:

```bash
cd tests
konduit --help
```

which should prompt all currently available commands, i.e.

```bash
Usage: konduit [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  build          Build the underlying konduit.jar (again).
  init           Initialize the konduit-python CLI.
  predict-numpy  Get predictions for your pipeline from numpy input.
  serve          Serve a pipeline from a konduit.yaml
  stop-server    Stop the Konduit server associated with a given YAML file.


```

For more help on individual commands you can do
```bash
konduit serve --help
```

to get help for the `serve` command (and all others in the same way)

```bash
Usage: konduit serve [OPTIONS]

  Serve a pipeline from a konduit.yaml

Options:
  --config TEXT        Relative or absolute path to your konduit serving YAML file.
  --start_server TEXT  Whether to start the server instance after initialization.
  --help               Show this message and exit.
```

## Basic usage

You should be in the `tests` folder still. There are a lot of YAML files in the `yaml` folder
to test Konduit with. Let's say we want to serve the pipeline described in `yaml/konduit.yaml`.
Let's have a look at it first:

```yaml
serving:
  http_port: 1337
  input_data_type: NUMPY
  output_data_type: NUMPY
  log_timings: True
  extra_start_args: -Xmx8g
  jar_path: konduit.jar
  sleep: 1
steps:
  python_step:
    type: PYTHON
    python_path: .
    python_code_path: ./simple.py
    python_inputs:
      first: NDARRAY
    python_outputs:
      second: NDARRAY
client:
    url: http://localhost:1337
```

This Konduit experiment describes how your model will be run, what inputs it takes and
what types of output it will generate. In essence, this configuration will run the script
`simple.py` on input data that you can specify. To serve this Konduit pipeline you can
just run:

```bash
konduit serve --config yaml/konduit.yaml
```

and to get predictions from it you can use:

```bash
konduit predict-numpy --config yaml/konduit.yaml --numpy_data ../data/input-0.npy 

```

### CODE REVIEW

File: python/cli.py [hunk 1]
```diff
@@ -69,10 +71,36 @@ def build(os):
     build_jar(os)
 
 
+@click.command()
+@click.option('--yaml', help='Relative or absolute path to your konduit serving YAML file.')
+@click.option('--start_server', default=True, help='Relative or absolute path to your konduit serving YAML file.')
```
Comment (ShamsUlAzeem) [inline python/cli.py]: Needs `start_server` description here

File: python/konduit_cli.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+```bash
+Usage: konduit-python serve [OPTIONS]
+
+  Serve a pipeline from a konduit.yaml
+
+Options:
+  --yaml TEXT          Relative or absolute path to your konduit serving YAML
+                       file.
+  --start_server TEXT  Relative or absolute path to your konduit serving YAML
```
Comment (ShamsUlAzeem) [inline python/konduit_cli.md]: here too

File: konduit-serving-api/pom.xml [hunk 3]
```diff
@@ -66,16 +66,16 @@
...
-                    <version>1.1</version>
-                </dependency>
             </dependencies>
         </profile>
     </profiles>
 
     <dependencies>
+        <dependency>
+            <groupId>com.googlecode.json-simple</groupId>
+            <artifactId>json-simple</artifactId>
```
Comment (agibsonccc) [inline konduit-serving-api/pom.xml]: We want json-simple gone. No reason for it to exist on the classpath. See: https://github.com/KonduitAI/konduit-serving/pull/62

File: python/konduit_cli.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+konduit-python --help
+```
+
+which should prompt all currently available commands, i.e.
+
+```bash
+Usage: konduit-python [OPTIONS] COMMAND [ARGS]...
+
+Options:
+  --help  Show this message and exit.
```
Comment (ShamsUlAzeem) [inline python/konduit_cli.md]: @maxpumperla can you extend this description to mention the ability to see help for individual sub command?

Comment (ShamsUlAzeem) [reply python/konduit_cli.md]: also we want to have some arguments as `required=True` such as the `--os` option for `build` and `init`

Comment (maxpumperla) [reply python/konduit_cli.md]: I do that below

**Comment by maxpumperla** [comment]
@agibsonccc @ShamsUlAzeem this is ready now, please review again

**Comment by ShamsUlAzeem** [comment]
@maxpumperla I saw that it needs a `konduit.jar` file all the time. If we don't have a konduit.jar file it won't show the `--help` details. Having it prompt something like, `build konduit-serving before proceeding with konduit build --os windows-x86_64` or something along the lines would be helpful since most users won't know that we're using a konduit jar file under the hood. Or maybe we can build a konduit jar file on startup?

**Pull request merged**